libraries/chain_kv/include/b1/session/bytes.hpp

```diff
@@ -0,0 +1,284 @@
+#pragma once
+
+#include <algorithm>
+#include <string_view>
+
+#include <boost/pool/pool.hpp>
+
+#include <b1/session/bytes_fwd_decl.hpp>
+#include <b1/session/key_value_fwd_decl.hpp>
+
+namespace eosio::session {
+
+// \brief An immutable type to represent a pointer and a length.
+//
+class bytes final {
+public:
+    friend class key_value;
+    
+    template <typename T, typename allocator>
+    friend bytes make_bytes(const T* data, size_t length, allocator& a);
```

Comment:
I had a few reasons that I went this route:
1.  At some point I wanted to minimize the scope of friends and I preferred friending a function than a class.
2.  I was debating implementing some sort of pool of this types.  I also felt this design would allow for performance enhancement opportunities.
3.  I liked the idea of keeping the creation logic separate from the type so we could allow injecting new overloaded free functions.

At this point though, we could make them constructors as well.